runtime/dict.go

```diff
@@ -755,6 +747,54 @@ func dictRepr(f *Frame, o *Object) (*Object, *BaseException) {
 	return NewStr(buf.String()).ToObject(), nil
 }
 
+func dictSetDefault(f *Frame, args Args, _ KWArgs) (*Object, *BaseException) {
+	argc := len(args)
+	if argc == 1 {
+		return nil, f.RaiseType(TypeErrorType, "setdefault expected at least 1 arguments, got 0")
+	}
+	if argc > 3 {
+		return nil, f.RaiseType(TypeErrorType, fmt.Sprintf("setdefault expected at most 2 arguments, got %v", argc-1))
+	}
+	expectedTypes := []*Type{DictType, ObjectType, ObjectType}
+	if argc == 2 {
+		expectedTypes = expectedTypes[:2]
+	}
+	if raised := checkMethodArgs(f, "setdefault", args, expectedTypes...); raised != nil {
+		return nil, raised
+	}
+	key := args[1]
+	d := toDictUnsafe(args[0])
+	hash, raised := Hash(f, key)
+	if raised != nil {
+		return nil, raised
+	}
+	d.mutex.Lock(f)
+	t := d.table
+	v := d.version
+	index, entry, raised := t.lookupEntry(f, hash.Value(), key)
+	var value *Object
+	if raised == nil {
+		if v != d.version {
+			// Dictionary was recursively modified. Blow up instead
+			// of trying to recover.
+			raised = f.RaiseType(RuntimeErrorType, "dictionary changed during write")
+		} else {
+			if entry == nil {
+				if argc > 2 {
+					value = args[2]
+				} else {
+					value = None
+				}
+				raised = addDictTableEntry(f, d, hash, key, value, index)
```

Comment:
Factoring this out is a good first step. I'm wondering though if we can go even further: do you think it's possible for putItem to take a function that will replace the table manipulation from line 409-426? I'm thinking that the callers that are deleting a key can pass a function that does this part:

```
if entry != nil && entry != deletedEntry {
	d.table.storeEntry(index, deletedEntry)
	d.table.incUsed(-1)
	d.incVersion()
}
```

While the callers that are setting a key can pass a function that will do this part:

```
newEntry := &dictEntry{hash.Value(), key, value}
if newTable, ok := t.writeEntry(f, index, newEntry); ok {
	if newTable != nil {
		d.storeTable(newTable)
	}
	d.incVersion()
} else {
	raised = f.RaiseType(OverflowErrorType, errResultTooLarge)
}
```

Finally, we could have a slightly different function for setdefault, something like:

```
if entry == nil || entry == deletedEntry {
	return
}
newEntry := &dictEntry{hash.Value(), key, value}
if newTable, ok := t.writeEntry(f, index, newEntry); ok {
	if newTable != nil {
		d.storeTable(newTable)
	}
	d.incVersion()
} else {
	raised = f.RaiseType(OverflowErrorType, errResultTooLarge)
}
```

The nice thing is that it encapsulates the locking, hashing and finding the entry. The function being called can then do its work in a safe, atomic way.

What do you think?